hal/src/common/thumbv6m/sercom/v2/spi.rs

```diff
@@ -0,0 +1,1768 @@
+    M: Mode,
+    C: CharSize,
+{
+    /// Reset the SERCOM peripheral
+    #[inline]
+    fn swrst(sercom: &P::Sercom) {
+        sercom.spi().ctrla.write(|w| w.swrst().set_bit());
+        while sercom.spi().syncbusy.read().swrst().bit_is_set() {}
+        unsafe { Self::reset_serial_read_state() };
+    }
+
+    /// Reset the `static` state used by the implementation of [`Read`]
+    ///
+    /// The state should be reset whenever the peripheral is reset or disabled.
+    /// It should not be possible for the state to become out of sync from safe
+    /// code.
+    unsafe fn reset_serial_read_state() {
+        SERIAL_READ_STATE[P::Sercom::NUM] = false;
+    }
+
+    /// Change the [`Config`] [`Mode`] or [`CharSize`]
+    #[inline]
+    fn change<M2, C2>(self) -> Config<P, M2, C2>
+    where
+        M2: Mode,
+        C2: CharSize,
+    {
+        Config {
+            sercom: self.sercom,
+            pads: self.pads,
+            mode: PhantomData,
+            chsize: PhantomData,
+            freq: self.freq,
+        }
+    }
+
+    /// Trigger the [`Sercom`]'s SWRST and return a [`Config`] in the
+    /// default configuration.
+    #[inline]
+    pub fn reset(self) -> Config<P> {
+        Config::create(self.sercom, self.pads, self.freq)
+    }
+
+    /// Obtain a reference to the PAC `SERCOM` struct
+    ///
+    /// Directly accessing the `SERCOM` could break the invariants of the
+    /// type-level tracking in this module, so it is unsafe.
+    #[inline]
+    pub unsafe fn sercom(&self) -> &P::Sercom {
+        &self.sercom
+    }
+
+    /// Consume the [`Config`], reset the peripheral, and return the [`Sercom`]
+    /// and [`Pads`]
+    #[inline]
+    pub fn free(self) -> (P::Sercom, P) {
+        Self::swrst(&self.sercom);
+        (self.sercom, self.pads)
+    }
+
+    /// Change the operating [`Mode`]
+    #[inline]
+    pub fn op_mode<M2: Mode>(self) -> Config<P, M2, C> {
+        M2::configure(&self.sercom);
+        self.change()
+    }
+
+    /// Change the [`CharSize`]
+    #[inline]
+    pub fn char_size<C2: CharSize>(self) -> Config<P, M, C2> {
+        C2::configure(&self.sercom);
+        self.change()
+    }
+
+    /// Change the clock polarity
+    #[inline]
+    pub fn cpol(self, cpol: spi::Polarity) -> Self {
+        let cpol = match cpol {
+            spi::Polarity::IdleLow => false,
+            spi::Polarity::IdleHigh => true,
+        };
+        self.sercom.spi().ctrla.modify(|_, w| w.cpol().bit(cpol));
+        self
+    }
+
+    /// Change the clock phase
+    #[inline]
+    pub fn cpha(self, cpha: spi::Phase) -> Self {
+        let cpha = match cpha {
+            spi::Phase::CaptureOnFirstTransition => false,
+            spi::Phase::CaptureOnSecondTransition => true,
+        };
+        self.sercom.spi().ctrla.modify(|_, w| w.cpha().bit(cpha));
+        self
+    }
+
+    /// Change the SPI mode (clock polarity & phase)
+    #[inline]
+    pub fn spi_mode(self, mode: spi::Mode) -> Self {
```

Comment:
`Mode`, `Phase`, `Polarity` are part of the API, could we re-export so peeps dont have to throw `use atsamd_hal::hal::spi::{Mode, Phase, Polarity};` everywhere?

Reply:
Good idea

Reply:
One problem with this. I define my own `Mode` trait in this module. I think I will skip re-exporting the embedded HAL `Mode`. If you're using the embedded HAL `Mode`, you're probably using one of the constants, e.g. `MODE_0`.

Reply:
Actually, I'll just export the `Mode` constants instead.